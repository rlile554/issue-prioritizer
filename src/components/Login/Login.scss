.keyForm {
	display: flex;
	flex-direction: column;
	align-items: center;
	color: rgb(107, 107, 122);
	width: 650px;
}

.title {
	font-size: 45px;
	font-weight: lighter;
}

.submitBtn {
	margin-top: 10px;
	width: 50%;
	border: none;
	background-color: inherit;
	font-size: 45px;
	font-weight: lighter;
	color: rgb(107, 107, 122);
	border: 1px solid #F8F8FF;
	border-radius: 15px;
}

.submitBtn:hover {
	border-color: rgb(107, 107, 122);
}

.keyInput {
	background-color: inherit;
	border: none;
	border-bottom: 3px solid rgb(107, 107, 122);
	font-size: 30px;
	width: 100%
}

.error {
	color: red;
}