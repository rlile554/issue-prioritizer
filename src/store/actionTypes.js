export const ADD_API_KEY = 'ADD_API_KEY';

export const ADD_REPOS = 'ADD_REPOS';
export const SELECT_REPO = 'SELECT_REPO';

export const ADD_ISSUES = 'ADD_ISSUES';
export const REORDER_ISSUES = 'REORDER_ISSUES';

export const LOADING = 'LOADING';