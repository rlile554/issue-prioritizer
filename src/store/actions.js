import {
	ADD_API_KEY,
	ADD_REPOS,
	LOADING,
	SELECT_REPO,
	ADD_ISSUES,
	REORDER_ISSUES,
} from './actionTypes';

export const addApiKey = (apiKey) => {
	return {
		type: ADD_API_KEY,
		apiKey,
	};
};

export const addRepos = (repos) => {
	return {
		type: ADD_REPOS,
		repos,
	};
};

export const selectRepo = (repoId) => {
	return {
		type: SELECT_REPO,
		repoId,
	};
};

export const addIssues = (issues, repoId) => {
	return {
		type: ADD_ISSUES,
		issues,
		repoId,
	};
};

export const reorderIssues = (repoId, sourceIndex, destinationIndex) => {
	return {
		type: REORDER_ISSUES,
		repoId,
		sourceIndex,
		destinationIndex
	}
}

export const loading = (isLoading) => {
	return {
		type: LOADING,
		isLoading,
	};
};
