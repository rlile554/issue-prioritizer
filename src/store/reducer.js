import {
	ADD_API_KEY,
	ADD_REPOS,
	LOADING,
	SELECT_REPO,
	ADD_ISSUES,
	REORDER_ISSUES,
} from './actionTypes';

const initialState = {
	loading: false,
	apiKey: '',
	repos: [],
	issuesList: [],
	selectedRepo: null,
};

const reducer = (state = initialState, action) => {
	switch (action.type) {
		case LOADING:
			return {
				...state,
				loading: action.isLoading,
			};
		case ADD_API_KEY:
			return {
				...state,
				apiKey: action.apiKey,
			};
		case ADD_REPOS:
			return {
				...state,
				repos: action.repos,
				loading: false,
			};
		case SELECT_REPO:
			return {
				...state,
				selectedRepo: action.repoId,
			};
		case ADD_ISSUES:
			const idx = state.issuesList.findIndex((i) => i.id === action.repoId);
			const issuesToAdd = { id: action.repoId, issues: action.issues};
			let newIssues;
			if (idx !== -1) {
				newIssues = [
					...state.issuesList.slice(0, idx),
					issuesToAdd,
					...state.issuesList.slice(idx + 1, state.issuesList.length),
				];
			} else {
				newIssues = [...state.issuesList, issuesToAdd];
			}
			return {
				...state,
				issuesList: newIssues,
			};
		case REORDER_ISSUES:
			const { sourceIndex, destinationIndex, repoId } = action;
			const returnArray = state.issuesList.map((issuesObject) => {
				if (issuesObject.id === repoId) {
					const newIssues = issuesObject.issues.slice();
					const issueToMove = newIssues.splice(sourceIndex, 1)[0];
					newIssues.splice(destinationIndex, 0, issueToMove);
					return {
						...issuesObject,
						issues: newIssues
					};
				}
				return issuesObject;
			});
			return {
				...state,
				issuesList: returnArray,
			};
		default:
			return state;
	}
};

export default reducer;
