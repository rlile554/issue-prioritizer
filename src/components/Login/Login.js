import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import './Login.scss';
import { addApiKey, loading } from '../../store/actions';
import { validateToken } from '../../utils/GitHubApi';

const Login = () => {
	const dispatch = useDispatch();
	const storedKey = useSelector((state) => (state.apiKey ? state.apiKey : ''));
	const [key, setKey] = useState(storedKey);
	const [error, setError] = useState(false);

	const handleChange = (e) => {
		setKey(e.target.value);
		setError(false);
	};

	const handleSubmit = async (e) => {
		e.preventDefault();
		dispatch(loading(true));

		try{
			await validateToken(key);
			dispatch(addApiKey(key));
			dispatch(loading(false));
		} catch(e) {
			setError(e.message)
		}
	};

	return (
		<div>
			<form className='keyForm' onSubmit={handleSubmit}>
				<h1 className='title'>Please enter your GitHub API key</h1>
				<input
					className='keyInput'
					placeholder='...'
					value={key}
					onChange={handleChange}
				/>
				{error && (
					<p className='error'>Invalid api key. Please insert a valid key</p>
				)}
				<button className='submitBtn' type='submit'>
					submit
				</button>
			</form>
		</div>
	);
};

export default Login;
