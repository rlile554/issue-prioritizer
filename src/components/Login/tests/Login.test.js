import React from 'react';
import Login from '../Login';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../../store/store';

test('Login renders', () => {
	const component = renderer.create(
		<Provider store={store}>
			<Login />
		</Provider>
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
