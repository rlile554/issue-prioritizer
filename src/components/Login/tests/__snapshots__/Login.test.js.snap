// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders 1`] = `
<div>
  <form
    className="keyForm"
    onSubmit={[Function]}
  >
    <h1
      className="title"
    >
      Please enter your GitHub API key
    </h1>
    <input
      className="keyInput"
      onChange={[Function]}
      placeholder="..."
      value=""
    />
    <button
      className="submitBtn"
      type="submit"
    >
      submit
    </button>
  </form>
</div>
`;
